plugins/in_dummy/in_dummy.c

```diff
@@ -274,22 +260,22 @@ static struct flb_config_map config_map[] = {
    },
    {
     FLB_CONFIG_MAP_STR, "dummy", DEFAULT_DUMMY_MESSAGE,
-    0, FLB_FALSE, 0,
+    0, FLB_TRUE, offsetof(struct flb_dummy, dummy_message),
```

Comment:
I have refactored the code to work properly by using flb_input_get_property. The code though is now extremely complex because the original code works in a very weird way where it uses the default when the parameter is invalid JSON.

```c
    /* handle it explicitly since we need to validate it is valid JSON */
    msg = flb_input_get_property("dummy", in);
    if (msg) {
        ret = flb_pack_json(msg, strlen(msg), &ctx->ref_msgpack,
                            &ctx->ref_msgpack_size, &root_type);
    }
    if (msg && ret == 0) {
    	ctx->dummy_message = flb_strdup(msg);
    	ctx->dummy_message_len = strlen(msg);
    } else  {
        flb_plg_warn(ctx->ins, "data is incomplete. Use default string.");

        ctx->dummy_message = flb_strdup(DEFAULT_DUMMY_MESSAGE);
        ctx->dummy_message_len = strlen(ctx->dummy_message);

        ret = flb_pack_json(ctx->dummy_message,
                            ctx->dummy_message_len,
                            &ctx->ref_msgpack, &ctx->ref_msgpack_size,
                            &root_type);
        if (ret != 0) {
            flb_plg_error(ctx->ins, "unexpected error");
            return -1;
        }
    }
```

I'm trying to think of a way to both support the original behavior and reduce the messy conditional complexity. Any suggestions?

Reply:
I refactored it to this for now:

```c
    /* handle it explicitly since we need to validate it is valid JSON */
    msg = flb_input_get_property("dummy", in);
    if (msg == NULL) {
        msg = DEFAULT_DUMMY_MESSAGE;
    }
    ret = flb_pack_json(msg, strlen(msg), &ctx->ref_msgpack,
                        &ctx->ref_msgpack_size, &root_type);
    if (ret == 0) {
        ctx->dummy_message = flb_strdup(msg);
        ctx->dummy_message_len = strlen(msg);
    } else {
        flb_plg_warn(ctx->ins, "data is incomplete. Use default string.");

        ctx->dummy_message = flb_strdup(DEFAULT_DUMMY_MESSAGE);
        ctx->dummy_message_len = strlen(ctx->dummy_message);

        ret = flb_pack_json(ctx->dummy_message,
                            ctx->dummy_message_len,
                            &ctx->ref_msgpack, &ctx->ref_msgpack_size,
                            &root_type);
        if (ret != 0) {
            flb_plg_error(ctx->ins, "unexpected error");
            return -1;
        }
    }
```

Still not entirely satisfied with it. If DEFAULT_DUMMY_MESSAGE is, for some weird reason, invalid JSON ti will parse it twice, but that should never happen in practice and that case is still handled.

Reply:
I'm going to check that I am not double freeing the dummy_message then marked this as resolved.